### PR TITLE
[`SFTTrainer`] Flash attention support for SFTTrainer

### PR DESCRIPTION
**Pull request by younesbelkada**
Related to https://github.com/huggingface/transformers/pull/25265

Users can easily benefit from flash attention, this PR adds a new argument in `SFTTrainer` to take care of that and properly document and raise errors when relevant. 

This leads to some interesting speedups and memory saving that I will detail after doing some experiments, hence putting this PR as draft for now

Only available if you use pytorch nightlies and use `packing=True` as SDPA + flash attention does not support padding

cc @lvwerra @fxmarty @vwxyzjn

### CODE REVIEW

**Comment by fxmarty** [comment]
Are trl users using padding?

**Comment by younesbelkada** [comment]
adding `packing=True` should lead to having no padding in the input texts as it will concatenate text chunks until reaching `max_seq_length`

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -29,6 +30,7 @@
 )
 from transformers.trainer_callback import TrainerCallback
 from transformers.trainer_utils import EvalPrediction
+from transformers.utils import ContextManagers
```
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:33]: I made this 😂

Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:33]: hahahah nice!

**Comment by younesbelkada** [comment]
Closing for now as https://github.com/huggingface/transformers/pull/25598 might be merged

**Pull request closed without merging**